src/main/java/org/broadinstitute/hellbender/engine/ReadShard.java

```diff
@@ -0,0 +1,182 @@
+package org.broadinstitute.hellbender.engine;
+
+import htsjdk.samtools.SAMSequenceDictionary;
+import htsjdk.samtools.util.Locatable;
+import org.broadinstitute.hellbender.engine.filters.ReadFilter;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.iterators.ReadFilteringIterator;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * A class to represent a shard of reads data, optionally expanded by a configurable amount of padded data.
+ *
+ * The reads are lazily loaded by default (when accessing the reads via {@link #iterator}. Loading all the
+ * reads in the window at once is possible via {@link #loadAllReads}.
+ *
+ * The reads returned will overlap the expanded padded interval. It's possible to query whether they are within
+ * the main part of the shard via {@link #isContainedWithinShard} and {@link #startsWithinShard}.
+ *
+ * The reads in the window can be filtered via {@link #setReadFilter} (no filtering is performed by default).
+ */
+public final class ReadShard implements Iterable<GATKRead>, Locatable {
```

Comment:
You raise a good point about `VariantShard`/`ReferenceShard` -- these are only used in an obsolete Spark implementation of BQSR that is slated for deletion, however (https://github.com/broadinstitute/gatk/issues/1203), so we will likely delete those two classes as part of that ticket.

I went with "shard" because "window" to me implies an interval rather than a packet of records, but I'm open to suggestions. Do you still dislike the name `ReadShard` even after learning that the existing `VariantShard`/`ReferenceShard` will likely be deleted?

(Naming is hard! :)  )